ESPS/general/man/man3/putsym.3
.\" Copyright (c) 1987 Entropic Speech, Inc.; All rights reserved
.\" @(#)putsym.3	1.7 21 Apr 1997 ESI
.TH PUTSYM 3\-ESPSu 21 Apr 1997
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech,, Inc.
.SH "NAME"
.nf
putsym_f, putsym_i, putsym_s - put a parameter value into an ESPS Common file
fputsym_f, fputsym_i, fputsym_s - put a parameter value into an arbitrary file
.fi
.SH "SYNOPSIS"
.B #include <stdio.h>
.br
.B #include <esps/esps.h>
.PP
.B int putsym_f (symbol, value)
.br
.B char	*symbol;
.br
.B float value;
.PP
.B int putsym_i (symbol, value)
.br
.B char	*symbol;
.br
.B int value;
.PP
.B int putsym_s (symbol, value)
.br
.B char	*symbol;
.br
.B char *value;
.PP
.B int fputsym_f (symbol, value, file)
.br
.B char	*symbol;
.br
.B float value;
.br
.B char *file;
.PP
.B int fputsym_i (symbol, value, file)
.br
.B char	*symbol;
.br
.B int value;
.br
.B char *file;
.PP
.B int fputsym_s (symbol, value, file)
.br
.B char	*symbol;
.br
.B char *value;
.br
.B char *file;
.SH "DESCRIPTION"
.PP
.I putsym_f,
.I putsym_i,
and
.I putsym_s,
all write the specified \fIsymbol\fP and its \fIvalue\fP
into an ESPS Common file if Common processing is enabled.
.PP
ESPS Common processing may be disabled by setting the environment
variable USE_ESPS_COMMON to "off".  The default ESPS Common file 
is .espscom in the user's home directory.  This may be overidden by setting
the environment variable ESPSCOM to the desired path.  
If Common processing is disabled, then these functions simply return
without doing anything.
.PP
.I fputsym_f,
.I fputsym_i,
and
.I fputsym_s,
all write the specified \fIsymbol\fP and its \fIvalue\fP into 
an arbitrary named file \fIfile\fP.  Their behavior does 
not depend on whether or not Common processing is enabled.  
.PP
In all cases, a return value of zero indicates success and a return
value of -1 indicates error (i.e. the Common file or specified file
could not be updated or created).
.PP
The output for all these functions is written in the same format as
ESPS Parameter files (see [1] for more details).  (Definite
assignments are used, without the optional prompt string and parameter
limitations).  Values written with
these functions can be read with \fIgetsym_xx\fP.
.SH "EXAMPLE"
.nf
/* parameter values to store in an ESPS Common file */
start = 5001;
dcrem = "no";

if (putsym_i ("start", start) != 0)
   (void) fprintf (stderr, "could not write into ESPS Common file.\n");
if (putsym_s ("dcrem", dcrem) != 0)
   (void) fprintf (stderr, "could not write into ESPS Common file.\n");

.fi
.SH DIAGNOSTICS
.PP
.SH "BUGS"
.PP
.SH "SEE ALSO"
.PP
\fIread_params\fP (3\-\s-1ESPS\s+1), \fIsymerr_exit\fP
(3\-\s-1ESPS\s+1), \fIgetsym\fP (3\-\s-1ESPS\s+1), \fIgetsymdef\fP
(3\-\s-1ESPS\s+1), \fIsymtype\fP (3\-\s-1ESPS\s+1), \fIsymsize\fP
(3\-\s-1ESPS\s+1), 
.br
\fIsymlist\fP (3\-\s-1ESPS\s+1), \fIsymdefinite\fP
(3\-\s-1ESPS\s+1), \fIsymchoices\fP (3\-\s-1ESPS\s+1), \fIsymrange\fP
(3\-\s-1ESPS\s+1), \fIsymprompt\fP (3\-\s-1ESPS\s+1),
\fIsymerr_exit\fP (3\-\s-1ESPS\s+1)
.SH "REFERENCES"
[1] ETM-S-86-12:jtb, Parameter Files in the Speech Processing System
.SH "AUTHOR"
Ajaipal S. Virdy, Alan Parker, John Shore


ESPS/general/src/lib/putsym.c
/*
  This material contains proprietary software of Entropic Speech, Inc.   
  Any reproduction, distribution, or publication without the prior	   
  written permission of Entropic Speech, Inc. is strictly prohibited.
  Any public distribution of copies of this work authorized in writing by
  Entropic Speech, Inc. must bear the notice			
 								
      "Copyright (c) 1987 Entropic Speech, Inc.; All rights reserved"

  This module contains the functions used to put symbols into the esps 
  common file.
*/
 				

#ifndef lint
	static char *sccs_id = "@(#)putsym.c	1.14	12/22/93 ESI";
#endif
 	

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <esps/esps.h>

#define	LINE_SIZE	256
#define	MAXLEN		256
#define	TRUE		1
#define	FALSE		0

#define	TMP_FILE	"/tmp/comXXXXXX"

char	*getenv(), *strcpy(), *strtok(), *mktemp();
char	*strcat();
void rewind();
double atof();

static FILE	*com_fp = NULL;		/* file pointer to ESPS Common file */
static char	filename[80];
static FILE	*tmp_fp;
static char	template[20];

static int
setup_common(file)
char *file;
{
char	*home;
char	*com;
struct stat combuf;

    com_fp = NULL;
    filename[0] = '\0';
    tmp_fp = NULL;
    strcpy(template,TMP_FILE);
    if(file == NULL) {
      if(getenv("USE_ESPS_COMMON") != NULL && 
         strcmp(getenv("USE_ESPS_COMMON"),"off") == 0)
	  return 0;

      if((com = getenv("ESPSCOM")) == NULL) {
        if( (home = getenv("HOME")) == NULL )
	  home = ".";
        (void) strcpy(filename,home);
        (void) strcat(filename,"/");
        (void) strcat(filename,SPS_COMMON);
       }
       else
        (void) strcpy(filename,com);
     } else {
       (void) strcpy(filename,file);
     }

    if ((com_fp = fopen (filename, "r")) == NULL) {
	if ((com_fp = fopen(filename, "w")) == NULL) {
	   (void) fprintf (stderr,
	   "putsym: could not create ESPS Common file: %s.\n",filename);
	   return -1;
	}
    } else {
	  if (stat(filename, &combuf) != 0) {
	     fprintf(stderr,"putsym: cannot stat common file %s\n",
		     filename);
	     fprintf(stderr,"This should not happen.\n");
	     exit (1);
	  }

	  if (combuf.st_mode & S_IFDIR) {
	    fprintf(stderr,
              "putsym: Warning, ESPS Common file %s is a directory.\n",
	      filename);
	    fprintf(stderr, "It is not being processed.\n");
	    fclose(com_fp);
	    return -1;
	  }
    }

    if ((tmp_fp = fopen (mktemp(template), "w+")) == NULL ) {
	(void) fprintf (stderr,
	"putsym: could not open temporary file %s for writing.\n",
	template);
	return -1;
    }
    return 1;
}


int
putsym_i (symbol, value)
char	*symbol;
int	value;
{



char	buffer[LINE_SIZE];	/* buffer to hold output text */
char	*text = buffer;		/* pointer to buffer */
char	*name = buffer;		/* string for holding name of symbol */
char	line[LINE_SIZE];
char	*type;
char	*val;

int	found = FALSE;
int     ret;

    spsassert(symbol != NULL, "putsym_i: symbol is NULL");

    if((ret = setup_common(NULL)) != 1) return ret;

/*
 * we've opened the file, now update it.
 * But first check if symbol already exists in the ESPS Common file.
 *
 */

#ifdef DEBUG
   (void) fprintf(stderr,"putsym_i: value: %d, symbol: %s\n",value,symbol);
#endif

     while (fgets(line, LINE_SIZE, com_fp) != NULL) {

	(void) strcpy (buffer, line);
 	if ( strcmp("int", (type = strtok(buffer, " \t"))) != 0 ) {
 
#ifdef DEBUG
   (void) fprintf (stderr, "putsym_i: skipping (%s) type\n", type);
#endif
	   (void) fputs (line,tmp_fp);
	   continue;

	}

	name = strtok(NULL, "\t =");
 	val = strtok(NULL, "\t ;");

#ifdef DEBUG
	(void) fprintf (stderr, "putsym_i: type: %s\n", type);
	(void) fprintf (stderr,	"          name: %s\n", name);
	(void) fprintf (stderr,	"         value: %d\n", atoi(val));
#endif


	if (strcmp(name, symbol) != 0) {
#ifdef DEBUG
	   (void) fprintf (stderr, "putsym_i: skipping (%s) symbol\n", name);
#endif
	   (void) fputs (line,tmp_fp);
	   continue;
	}

#ifdef DEBUG
	(void) fprintf (stderr, "putsym_i: updating (%s) symbol\n", name);
#endif

	(void)sprintf (text, "int %s = %d;\n", symbol, value);
	(void) fputs (text,tmp_fp);
	found = TRUE;

     } /* end while */

     if ( !found ) {
	/* Looks like the symbol didn't exist in the ESPS Common file
	 * so let's append it to the end of the file. */
#ifdef DEBUG
	(void) fprintf (stderr, "putsym_i: new symbol (%s)\n", name);
#endif
	(void)sprintf (text, "int %s = %d;\n", symbol, value);
	(void) fputs (text,tmp_fp);
     }

/*
 * Now copy contents of TMP_FILE to ESPS Common file,
 * but rewind both files first
 *
 */
    (void) rewind (tmp_fp);
    (void) fclose (com_fp);

    if ((com_fp = fopen (filename, "w")) == NULL) {
	(void) fprintf (stderr,
	"putsym_i: could not open ESPS Common file %s for writing.\n",
	filename);
	return -1;
    }

    while (fgets(line, LINE_SIZE, tmp_fp) != NULL) {
	(void)fprintf (com_fp, "%s", line);
#ifdef DEBUG
	(void)fprintf (stderr, "%s", line);
#endif
    }

    (void) fclose (tmp_fp);
    (void) fclose (com_fp);
    (void) unlink (template);

    return 0;

} /* end putsym_i() */

int
fputsym_i (symbol, value, file)
char	*symbol;
int	value;
char	*file;
{

char	buffer[LINE_SIZE];	/* buffer to hold output text */
char	*text = buffer;		/* pointer to buffer */
char	*name = buffer;		/* string for holding name of symbol */
char	line[LINE_SIZE];


char	*type;
char	*val;

int	found = FALSE;
int     ret;

    spsassert(symbol != NULL, "putsym_i: symbol is NULL");
    spsassert(file, "fputsym_i: filename is NULL");

    if((ret = setup_common(file)) != 1) return ret;
/*
 * we've opened the file, now update it.
 * But first check if symbol already exists in the ESPS Common file.
 *
 */

#ifdef DEBUG
   (void) fprintf(stderr,"fputsym_i: value: %d, symbol: %s\n",value,symbol);
#endif

     while (fgets(line, LINE_SIZE, com_fp) != NULL) {

	(void) strcpy (buffer, line);
 	if ( strcmp("int", (type = strtok(buffer, " \t"))) != 0 ) {
 
#ifdef DEBUG
   (void) fprintf (stderr, "fputsym_i: skipping (%s) type\n", type);
#endif
	   (void) fputs (line,tmp_fp);
	   continue;

	}

	name = strtok(NULL, "\t =");
 	val = strtok(NULL, "\t ;");

#ifdef DEBUG
	(void) fprintf (stderr, "fputsym_i: type: %s\n", type);
	(void) fprintf (stderr,	"          name: %s\n", name);
	(void) fprintf (stderr,	"         value: %d\n", atoi(val));
#endif


	if (strcmp(name, symbol) != 0) {
#ifdef DEBUG
	   (void) fprintf (stderr, "fputsym_i: skipping (%s) symbol\n", name);
#endif
	   (void) fputs (line,tmp_fp);
	   continue;
	}

#ifdef DEBUG
	(void) fprintf (stderr, "fputsym_i: updating (%s) symbol\n", name);
#endif

	(void)sprintf (text, "int %s = %d;\n", symbol, value);
	(void) fputs (text,tmp_fp);
	found = TRUE;

     } /* end while */

     if ( !found ) {
	/* Looks like the symbol didn't exist in the ESPS Common file
	 * so let's append it to the end of the file. */
#ifdef DEBUG
	(void) fprintf (stderr, "putsym_i: new symbol (%s)\n", name);
#endif
	(void)sprintf (text, "int %s = %d;\n", symbol, value);
	(void) fputs (text,tmp_fp);
     }

/*
 * Now copy contents of TMP_FILE to ESPS Common file,
 * but rewind both files first
 *
 */
    (void) rewind (tmp_fp);
    (void) fclose (com_fp);

    if ((com_fp = fopen (filename, "w")) == NULL) {
	(void) fprintf (stderr,
	"fputsym_i: could not open file %s for writing.\n",
	filename);
	return -1;
    }

    while (fgets(line, LINE_SIZE, tmp_fp) != NULL) {
	(void)fprintf (com_fp, "%s", line);
#ifdef DEBUG
	(void)fprintf (stderr, "%s", line);
#endif
    }

    (void) fclose (tmp_fp);
    (void) fclose (com_fp);
    (void) unlink (template);

    return 0;

} /* end putsym_i() */



int
fputsym_s (symbol, value, file)
char	*symbol;
char	*value;
char	*file;
{

char	buffer[LINE_SIZE];	/* buffer to hold output text */
char	*text = buffer;		/* pointer to buffer */
char	*name = buffer;		/* string for holding name of symbol */
char	line[LINE_SIZE];


char	*type;
char	*val;
int	found = FALSE;
int     ret;

char	DQUOTE = '"';
    spsassert(symbol != NULL, "fputsym_s: symbol is NULL");
    spsassert(file, "fputsym_s: filename is NULL");

    if((ret = setup_common(file)) != 1) return ret;

/*
 * we've opened the file, now update it.
 * But first check if symbol already exists in the ESPS Common file.
 *
 */

#ifdef DEBUG
    (void) fprintf (stderr,
    "\nfputsym_s: ESPS Common file %s has been opened for updating.\n",
    filename);
#endif

     while (fgets(line, LINE_SIZE, com_fp) != NULL) {

	(void) strcpy (buffer, line);
 	if ( strcmp("string", (type = strtok(buffer, " \t"))) != 0 ) {
 
#ifdef DEBUG
   (void) fprintf (stderr, "fputsym_s: skipping (%s) type\n", type);
#endif
	   (void)fprintf (tmp_fp, "%s", line);
	   continue;

	}

	name = strtok(NULL, "\t =");
 	val = strtok(NULL, "\t ;");

#ifdef DEBUG
	(void) fprintf (stderr, "fputsym_s: type: %s\n", type);
	(void) fprintf (stderr,	"          name: %s\n", name);
	(void) fprintf (stderr,	"         value: %s\n", val);
#endif


	if ( strcmp(name, symbol) != 0) {
#ifdef DEBUG
	   (void) fprintf (stderr, "fputsym_s: skipping (%s) symbol\n", name);
#endif
	   (void) fprintf (tmp_fp, "%s", line);
	   continue;
	}

#ifdef DEBUG
	(void) fprintf (stderr, "fputsym_s: updating (%s) symbol\n", name);
#endif

	(void)sprintf (text, "string %s = %c%s%c;\n", symbol, DQUOTE, value, DQUOTE);
	(void) fprintf (tmp_fp, "%s", text);
	found = TRUE;

     } /* end while */

     if ( !found ) {
	/* Looks like the symbol didn't exist in the ESPS Common file
	 * so let's append it to the end of the file. */
#ifdef DEBUG
	(void) fprintf (stderr, "fputsym_s: new symbol (%s)\n", name);
#endif
	(void)sprintf (text, "string %s = %c%s%c;\n", symbol, DQUOTE, value, DQUOTE);
	(void) fprintf (tmp_fp, "%s", text);
     }

/*
 * Now copy contents of TMP_FILE to ESPS Common file,
 * but rewind both files first
 *
 */
    (void) rewind (tmp_fp);
    (void) fclose (com_fp);

    if ((com_fp = fopen (filename, "w")) == NULL) {
	(void) fprintf (stderr,
	"fputsym_s: could not open file %s for writing.\n",
	filename);
	return -1;
    }

    while (fgets(line, LINE_SIZE, tmp_fp) != NULL) {
	(void)fprintf (com_fp, "%s", line);
#ifdef DEBUG
	(void)fprintf (stderr, "%s", line);
#endif
    }

    (void) fclose (tmp_fp);
    (void) fclose (com_fp);
    (void) unlink (template);

    return 0;

} /* end fputsym_s() */

int
putsym_s (symbol, value)
char	*symbol;
char	*value;
{

char	buffer[LINE_SIZE];	/* buffer to hold output text */
char	*text = buffer;		/* pointer to buffer */
char	*name = buffer;		/* string for holding name of symbol */
char	line[LINE_SIZE];


char	*type;
char	*val;
int	found = FALSE;
int     ret;

char	DQUOTE = '"';
    spsassert(symbol != NULL, "putsym_s: symbol is NULL");

    if((ret = setup_common(NULL)) != 1) return ret;


/*
 * we've opened the file, now update it.
 * But first check if symbol already exists in the ESPS Common file.
 *
 */

#ifdef DEBUG
    (void) fprintf (stderr,
    "\nputsym_s: ESPS Common file %s has been opened for updating.\n",
    filename);
#endif

     while (fgets(line, LINE_SIZE, com_fp) != NULL) {

	(void) strcpy (buffer, line);
 	if ( strcmp("string", (type = strtok(buffer, " \t"))) != 0 ) {
 
#ifdef DEBUG
   (void) fprintf (stderr, "putsym_s: skipping (%s) type\n", type);
#endif
	   (void)fprintf (tmp_fp, "%s", line);
	   continue;

	}

	name = strtok(NULL, "\t =");
 	val = strtok(NULL, "\t ;");

#ifdef DEBUG
	(void) fprintf (stderr, "putsym_s: type: %s\n", type);
	(void) fprintf (stderr,	"          name: %s\n", name);
	(void) fprintf (stderr,	"         value: %s\n", val);
#endif


	if ( strcmp(name, symbol) != 0) {
#ifdef DEBUG
	   (void) fprintf (stderr, "putsym_s: skipping (%s) symbol\n", name);
#endif
	   (void) fprintf (tmp_fp, "%s", line);
	   continue;
	}

#ifdef DEBUG
	(void) fprintf (stderr, "putsym_s: updating (%s) symbol\n", name);
#endif

	(void)sprintf (text, "string %s = %c%s%c;\n", symbol, DQUOTE, value, DQUOTE);
	(void) fprintf (tmp_fp, "%s", text);
	found = TRUE;

     } /* end while */

     if ( !found ) {
	/* Looks like the symbol didn't exist in the ESPS Common file
	 * so let's append it to the end of the file. */
#ifdef DEBUG
	(void) fprintf (stderr, "putsym_s: new symbol (%s)\n", name);
#endif
	(void)sprintf (text, "string %s = %c%s%c;\n", symbol, DQUOTE, value, DQUOTE);
	(void) fprintf (tmp_fp, "%s", text);
     }

/*
 * Now copy contents of TMP_FILE to ESPS Common file,
 * but rewind both files first
 *
 */
    (void) rewind (tmp_fp);
    (void) fclose (com_fp);

    if ((com_fp = fopen (filename, "w")) == NULL) {
	(void) fprintf (stderr,
	"putsym_s: could not open ESPS Common file %s for writing.\n",
	filename);
	return -1;
    }

    while (fgets(line, LINE_SIZE, tmp_fp) != NULL) {
	(void)fprintf (com_fp, "%s", line);
#ifdef DEBUG
	(void)fprintf (stderr, "%s", line);
#endif
    }

    (void) fclose (tmp_fp);
    (void) fclose (com_fp);
    (void) unlink (template);

    return 0;

} /* end putsym_s() */


int
putsym_f (symbol, value)
char	*symbol;
float	value;
{

char	buffer[LINE_SIZE];	/* buffer to hold output text */
char	*text = buffer;		/* pointer to buffer */
char	*name = buffer;		/* string for holding name of symbol */
char	line[LINE_SIZE];


char	*type;
char	*val;

int	found = FALSE;
int     ret;

    spsassert(symbol != NULL, "putsym_f: symbol is NULL");
    if((ret = setup_common(NULL)) != 1) return ret;

/*
 * we've opened the file, now update it.
 * But first check if symbol already exists in the ESPS Common file.
 *
 */

#ifdef DEBUG
   (void) fprintf(stderr,"putsym_f: value: %f, symbol: %s\n",value,symbol);
#endif

     while (fgets(line, LINE_SIZE, com_fp) != NULL) {

	(void) strcpy (buffer, line);
 	if ( strcmp("float", (type = strtok(buffer, " \t"))) != 0 ) {
 
#ifdef DEBUG
   (void) fprintf (stderr, "putsym_f: skipping (%s) type\n", type);
#endif
	   (void) fputs (line,tmp_fp);
	   continue;

	}

	name = strtok(NULL, "\t =");
 	val = strtok(NULL, "\t ;");

#ifdef DEBUG
	(void) fprintf (stderr, "putsym_f: val (string): %s\n",val);
	(void) fprintf (stderr, "putsym_f: type: %s\n", type);
	(void) fprintf (stderr,	"          name: %s\n", name);
	(void) fprintf (stderr,	"         value: %f\n", (float)atof(val));
#endif


	if (strcmp(name, symbol) != 0) {
#ifdef DEBUG
	   (void) fprintf (stderr, "putsym_f: skipping (%s) symbol\n", name);
#endif
	   (void) fputs (line,tmp_fp);
	   continue;
	}

#ifdef DEBUG
	(void) fprintf (stderr, "putsym_f: updating (%s) symbol\n", name);
#endif

	(void)sprintf (text, "float %s = %f;\n", symbol, value);
	(void) fputs (text,tmp_fp);
#ifdef DEBUG
        (void) fprintf(stderr,"text: %s\n",text);
#endif
	found = TRUE;

     } /* end while */

     if ( !found ) {
	/* Looks like the symbol didn't exist in the ESPS Common file
	 * so let's append it to the end of the file. */
#ifdef DEBUG
	(void) fprintf (stderr, "putsym_f: new symbol (%s)\n", name);
#endif
	(void)sprintf (text, "float %s = %f;\n", symbol, value);
	(void) fputs (text,tmp_fp);
     }

/*
 * Now copy contents of TMP_FILE to ESPS Common file,
 * but rewind both files first
 *
 */
    (void) rewind (tmp_fp);
    (void) fclose (com_fp);

    if ((com_fp = fopen (filename, "w")) == NULL) {
	(void) fprintf (stderr,
	"putsym_f: could not open ESPS Common file %s for writing.\n",
	filename);
	return -1;
    }

    while (fgets(line, LINE_SIZE, tmp_fp) != NULL) {
	(void)fprintf (com_fp, "%s", line);
#ifdef DEBUG
	(void)fprintf (stderr, "%s", line);
#endif
    }

    (void) fclose (tmp_fp);
    (void) fclose (com_fp);
    (void) unlink (template);

    return 0;

} /* end putsym_f() */


int
fputsym_f (symbol, value, file)
char	*symbol;
float	value;
char	*file;
{


char	buffer[LINE_SIZE];	/* buffer to hold output text */
char	*text = buffer;		/* pointer to buffer */
char	*name = buffer;		/* string for holding name of symbol */
char	line[LINE_SIZE];


char	*type;
char	*val;

int	found = FALSE;
int     ret;

    spsassert(symbol != NULL, "fputsym_f: symbol is NULL");
    spsassert(file, "fputsym_f: filename is NULL");

    if((ret = setup_common(file)) != 1) return ret;


/*
 * we've opened the file, now update it.
 * But first check if symbol already exists in the ESPS Common file.
 *
 */

#ifdef DEBUG
   (void) fprintf(stderr,"fputsym_f: value: %f, symbol: %s\n",value,symbol);
#endif

     while (fgets(line, LINE_SIZE, com_fp) != NULL) {

	(void) strcpy (buffer, line);
 	if ( strcmp("float", (type = strtok(buffer, " \t"))) != 0 ) {
 
#ifdef DEBUG
   (void) fprintf (stderr, "putsym_f: skipping (%s) type\n", type);
#endif
	   (void) fputs (line,tmp_fp);
	   continue;

	}

	name = strtok(NULL, "\t =");
 	val = strtok(NULL, "\t ;");

#ifdef DEBUG
	(void) fprintf (stderr, "val (string): %s\n", val);
	(void) fprintf (stderr, "putsym_f: type: %s\n", type);
	(void) fprintf (stderr,	"          name: %s\n", name);
	(void) fprintf (stderr,	"         value: %f\n", (float)atof(val));
#endif


	if (strcmp(name, symbol) != 0) {
#ifdef DEBUG
	   (void) fprintf (stderr, "fputsym_f: skipping (%s) symbol\n", name);
#endif
	   (void) fputs (line,tmp_fp);
	   continue;
	}

#ifdef DEBUG
	(void) fprintf (stderr, "putsym_f: updating (%s) symbol\n", name);
#endif

	(void)sprintf (text, "float %s = %f;\n", symbol, value);
#ifdef DEBUG
        (void) fprintf(stderr,"text: %s\n",text);
#endif
	(void) fputs (text,tmp_fp);
	found = TRUE;

     } /* end while */

     if ( !found ) {
	/* Looks like the symbol didn't exist in the ESPS Common file
	 * so let's append it to the end of the file. */
#ifdef DEBUG
	(void) fprintf (stderr, "putsym_f: new symbol (%s)\n", name);
#endif
	(void)sprintf (text, "float %s = %f;\n", symbol, value);
	(void) fputs (text,tmp_fp);
     }

/*
 * Now copy contents of TMP_FILE to ESPS Common file,
 * but rewind both files first
 *
 */
    (void) rewind (tmp_fp);
    (void) fclose (com_fp);

    if ((com_fp = fopen (filename, "w")) == NULL) {
	(void) fprintf (stderr,
	"putsym_f: could not open ESPS Common file %s for writing.\n",
	filename);
	return -1;
    }

    while (fgets(line, LINE_SIZE, tmp_fp) != NULL) {
	(void)fprintf (com_fp, "%s", line);
#ifdef DEBUG
	(void)fprintf (stderr, "%s", line);
#endif
    }

    (void) fclose (tmp_fp);
    (void) fclose (com_fp);
    (void) unlink (template);

    return 0;

} /* end putsym_f() */
